torch/csrc/distributed/c10d/ProcessGroup.cpp

```diff
@@ -158,3 +159,149 @@ void ProcessGroup::release_resources() {
+    auto it = registry_.find(storage);
+    if (it == registry_.end()) {
+      return {};
+    }
+    auto works = it->second;
+    registry_.erase(it);
+    return works;
+  }
+
+  void unregister_completed_works() {
+    std::unique_lock lock(lock_);
+    for (auto it = registry_.begin(); it != registry_.end();) {
+      std::vector<c10::intrusive_ptr<c10d::Work>> uncompleted_works;
+      for (const auto& work : it->second) {
+        if (work.defined()) {
+          if (work->canCheckIsCompleted()) {
+            if (!work->isCompleted()) {
+              uncompleted_works.push_back(work);
+            }
+          } else {
+            // if isCompleted() cannot be checked, we assume the work is not
+            // completed.
+            uncompleted_works.push_back(work);
+          }
+        }
+      }
+      if (uncompleted_works.empty()) {
+        it = registry_.erase(it);
+      } else {
+        it->second = std::move(uncompleted_works);
+        ++it;
+      }
+    }
+  }
+
+  size_t get_work_registry_size() {
+    std::unique_lock lock(lock_);
+    size_t total_size = 0;
+    for (const auto& [storage, works] : registry_) {
+      total_size += works.size();
+    }
+    return total_size;
+  }
+
+  ~WorkRegistry() {
+    // If there are still unwaited work objects, their corresponding process
+    // groups should have already been destroyed at this stage. Any attempts to
+    // wait for these work objects or to destroy them will only result in
+    // confusing errors. Therefore, we simply issue a warning and intentionally
+    // allow the unwaited work objects to leak.
+    size_t registry_size = get_work_registry_size();
+    if (registry_size > 0) {
+      TORCH_WARN(
+          "At the time of process termination, there are still ",
+          registry_size,
+          " unwaited collective calls. "
+          "Please review your program to ensure that:\n"
+          "1. c10d_functional.wait_tensor() is invoked on all tensors returned from c10d_functional collective,\n"
+          "2. work.wait() is invoked on work object returned from torch.distributed collective with async_op=True,\n"
+          "before the output tensors of the collective are used.");
+    }
+    for (auto& it : registry_) {
+      for (auto& work : it.second) {
+        work.release();
+      }
+    }
+  }
+
+ private:
+  std::unordered_map<
+      c10::weak_intrusive_ptr<c10::StorageImpl>,
+      std::vector<c10::intrusive_ptr<c10d::Work>>>
+      registry_;
+  std::mutex lock_;
+};
+
+static WorkRegistry process_registry;
+
+} // namespace
+
+namespace c10d {
+
+void register_work(
+    const at::Tensor& tensor,
+    const c10::intrusive_ptr<c10d::Work>& work) {
+  // Always clean up previously completed work objects, so that even if
+  // the user keeps issuing new collectives without waiting on previous ones,
+  // the registry size would not grow unbounded.
+  RankLocal<WorkRegistry>::get().unregister_completed_works();
```

Comment:
Hmm do we need/want to go through the registry everytime a new work is registered? Also it seems that the cleanup wouldn't happen for the last work entry in the program?

As discussed, I really think the entries should be evicted by `Work::wait()` on a per work basis. Do you think there's a problem with that approach?

Reply:
> As discussed, I really think the entries should be evicted by Work::wait() on a per work basis. Do you think there's a problem with that approach?

After reading more, it seems that we now have a one-to-many relationship from storage -> work. Eviction on a per-work basis would be more complex now, but I still think we should do it, because otherwise the registry is keeping the work alive for longer.

I suggesting something like `WorkRegistry::unregister_work(const c10::intrusive_ptr<c10d::Work>& work)` and call it in `Work::wait()`. We go through the registry and remove the **specified** work from the list associated with every storage. If a list becomes empty, we remove the entry from the registry.

Also there are some potential issues with your current impl: (1) not all work implementation has `checkCompleted` (2) `checkCompleted` can be quite expensive (`WorkNCCL::checkCompleted` does a bunch of things and an event query: https://fburl.com/code/qu5hi37w, doing these for all active work whenever a new work is registered may cause some unwanted effects). The suggestion above should resolve these issues because it only checks the identity of the work.

Reply:
sounds great - changed to call `unregister_work()` in `Work::wait()`